### ISSUE TITLE
Security vulnerabilities in v9.0.0 dependencies

### ISSUE THREAD

**Issue by kachkaev** [spencermountain/wtf_wikipedia#431]
👋  @spencermountain, thanks for this library! I got from _how to I parse wikitext templates?_ to _problem solved!_ in less than 30 mins – great stuff 💯

After pushing my code to GitHub I noticed two security advisories and decided to open this issue. Here’s what I see:

```sh
yarn why wtf_wikipedia
```
```txt
yarn why v1.22.10
[1/4] 🤔  Why do we have the module "wtf_wikipedia"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "wtf_wikipedia@9.0.0"
info Has been hoisted to "wtf_wikipedia"
info This module exists because it's specified in "dependencies".
info Disk size without dependencies: "2.27MB"
info Disk size with unique dependencies: "2.86MB"
info Disk size with transitive dependencies: "7.25MB"
info Number of shared dependencies: 52
```

---

```sh
yarn audit      
```

```txt
yarn audit v1.22.10
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ low           │ Prototype Pollution                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ yargs-parser                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=13.1.2 <14.0.0 || >=15.0.1 <16.0.0 || >=18.1.2             │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ wtf_wikipedia                                                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ wtf_wikipedia > recursive-install > yargs > yargs-parser     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1500                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ high          │ Regular Expression Denial of Service                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ trim-newlines                                                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=3.0.1 <4.0.0 || >=4.0.1                                    │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ wtf_wikipedia                                                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ wtf_wikipedia > path-exists-cli > meow > trim-newlines       │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1753                        │
└───────────────┴──────────────────────────────────────────────────────────────┘
2 vulnerabilities found - Packages audited: 1354
Severity: 1 Low | 1 High
```

Would you accept a PR that upgrades or removes vulnerable dependencies? 🙂 Thanks for the lib again! 🙌

**Comment by spencermountain** [comment]
Of course! Thank you Alexander.

**Comment by wvanderp** [comment]
I was looking at this because I tought it was a easy fix. but the problem is currently 4 levels deep and just updating ower dependincies is not enough. (altough there are some updates are not out of place).

so the bulk of the work here is getting the dependiesies to update. 

I doest help that i've choisen dependisies that aren't maintained any more.

**Comment by kachkaev** [comment]
The situation in v9.0.1 is better, there is only this:

```
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ moderate      │ Prototype Pollution in yargs-parser                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ yargs-parser                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=5.0.1                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ wtf_wikipedia                                                │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ wtf_wikipedia > recursive-install > yargs > yargs-parser     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://www.npmjs.com/advisories/1002557                     │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

You are right @wvanderp – this won’t be an easy fix. `yargs-parser` v3 is introduced via `recursive-install` which was [last published](https://www.npmjs.com/package/recursive-install?activeTab=versions) in June 2018. I was able to silence `yarn audit` by adding this section to `package.json` and reinstalling dependencies.

```json
  "resolutions": {
    "wtf_wikipedia/recursive-install/yargs": "^17.2.1",
  },
```

Looking at how `recursive-install` is used, I wonder if it should be moved from `"dependencies"` to `devDependencies"`? This might be a permanent fix for library users.

**Comment by wvanderp** [comment]
A progress report on this issue:

I've created a PR for recursive-install. If that gets adopted, most of the problems are solved.
I've also updated other dependencies of this project.
I've implemented the devDependencies mitigation in #443

some future solutions are:
using `resolutions` for the yarn users. but this is not supported by npm
switching over to my fork 
implementing our own

**Comment by spencermountain** [comment]
thank you wouter

**Comment by spencermountain** [comment]
fixed in 9.0.2
cheers
